### DISCUSSION TITLE
Personal/owned app instances

### DISCUSSION THREAD

**Discussion by nblumhardt** [datalust/seq-tickets#1095] [Ideas]
Currently, non-administrative users can send notifications to Seq Apps (e.g. the Slack or Email notifiers), but only administrators can stream events to an app.

This makes some integration and alerting scenarios awkward - for example, a user may wish to receive the full payloads of error events via email, or, record some stream of useful information to an external tool for analysis.

We've historically steered away from allowing non-administrative users to configure apps at all, but in Seq 4.2 we introduced the concept of overridable settings on app instances. This enables the currently-supported notification system by limiting the settings that non-admins can configure to a predetermined set.

If we were to allow non admins to create personal "outputs" that effectively behaved like app instances, but only using pre-installed apps and whitelisted properties, these could fill this feature gap.

It's also possible that the concept of an "output" might be useful for administrative users, too, since it would separate the configuration of mundane "global" properties (such as SMTP server details) associated with app instances, from the use-case-specific properties (such as "Subject" or "To" or "Channel") associated with individual outputs.

**Comment by nblumhardt** [comment]
We're making some moves towards this in 2020.1 and will revisit this in the future when we've introduced some new abstractions to support it.

**Comment by tbolon** [comment]
I would love to see this feature. Let me share our use case:

I want to allow other developers from our team to add notifications easily when a certain type of event occurs in seq. Previously they added custom code on the app to send an email to their inbox. I want to get ride of the custom code in our app and get ride of the email.

I now suggest them to replace this custom code with a specific log message sent to seq via serilog, and subscribe to this event type, and attach an app which send the seq message to MS Teams on a specific channel, or still send an email. It's their choice and they could decide to change this later. They do not know seq very well and I must explain them how to configure that.

For now they have two possibilities:

- They can create a dashboard and a graph for this specific event type with an alert. Pros: they just have to choose an existing generic app instance. Cons: they are not interested in the graph or the dashboard, and only want to be notified. Moreover, the Teams app currently has a limitation where a dashboard alert notification only contains a link to a query where the events are aggregated. They can't easily browse the raised events. I plan on submitting a PR to address that (by adding a link to the events in the teams notification, or add a setting to change the link attached to the teams notification.

- The can create a new app instance filtered on a signal they have created before. Cons: they have to create a signal and add a new app instance, copy all settings from another existing instance (webhook url). And it will result in another instance of the app on the server (adding a new AppDomain and increasing the memory consumption, if I read the settings well), increasing the workload.

There are multiple ways to handle this scenario, and I think you have already identified the problems with app instances and alerts.
The RFC #642 could help for that, but I am not entirely sure. 
I just wanted to share our daily usage and the problem we want to address with seq.

**Comment by nblumhardt** [comment]
Thanks @tbolon - this is great feedback to have. Some more recent work is driving towards this, too. I'll reopen so that we have the opportunity to discuss this as a team post 2020.1.
